### ISSUE TITLE
torrent_handle::status() hangs up on a paused torrent

### ISSUE THREAD

**Issue by evgenyzinoviev** [arvidn/libtorrent#134]
I have probably similar problem as described here: https://code.google.com/p/libtorrent/issues/detail?id=360
I use libtorrent version 1.0.6, boost 1.56.0 on linux x86_64.

First I add a torrent with `auto_managed` set to `false`. Then I `pause()` it, and then I print it's status every second:

```
    torrent_status status;
    do {
        status = handle.status(0);

        printf("%.2f%% (down: %.1f kb/s, up: %.1f kb/s, peers: %d, paused: %d, auto_managed: %d, qp: %d) %s\n",
               status.progress * 100,
               (double) status.download_rate / 1000,
               (double) status.upload_rate / 1000,
               status.num_peers,
               status.paused ? 1 : 0,
               status.auto_managed ? 1 : 0,
               status.queue_position,
               torrent_handle_state_str[status.state]);
        ::sleep(1);
    } while (true);
```

The first 6 iterations are ok, but on 7th iteration handle.status() hangs up, CPU usage grows up to 100%, and this is gdb backtrace:

```
(gdb) bt
#0  0x00007ffff749854f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff7ab3bc3 in libtorrent::torrent_handle::status(unsigned int) const () from /usr/lib64/libtorrent-rasterbar.so.8
#2  0x0000000000480ecd in simple_torrent_client (path="/home/user/Downloads/test.torrent") at /home/user/dev/test_torrent/torrent.cpp:432
#3  0x000000000046f283 in main (argc=1, argv=0x7fffffffd958) at /home/user/dev/test_torrent/main.cpp:66
```

The same thing happens if I set `auto_managed` to `true`, but prioritize all files to 0.

Basically, I just want to start it without any file being downloaded, and if I don't use `torrent_handle::status()`, everything is ok, but I want to get some information about a paused torrent. What am I doing wrong, or is it a libtorrent's bug, and if yes, how can I work around this?

Thank you!

**Comment by arvidn** [comment]
would you mind printing the stack of the thread that's stuck? i.e. when this happens, break and issue the command:

`thread apply all bt`

**Comment by evgenyzinoviev** [comment]
```
(gdb) thread apply all bt

Thread 3 (Thread 0x7ffff59fd700 (LWP 22346)):
#0  0x00007ffff749854f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff794c386 in ?? () from /usr/lib64/libtorrent-rasterbar.so.8
#2  0x0000000000464354 in boost::asio::detail::boost_asio_detail_posix_thread_function (arg=0x6db790)
    at /usr/include/boost/asio/detail/impl/posix_thread.ipp:64
#3  0x00007ffff7494204 in start_thread () from /lib64/libpthread.so.0
#4  0x00007ffff67b178d in clone () from /lib64/libc.so.6

Thread 2 (Thread 0x7ffff51fc700 (LWP 22347)):
#0  0x000000000046b58d in boost::asio::detail::task_io_service::reset (this=0x6da060)
    at /usr/include/boost/asio/detail/impl/task_io_service.ipp:247
#1  0x00000000004604c4 in boost::asio::io_service::reset (this=0x6d8b18)
    at /usr/include/boost/asio/impl/io_service.ipp:120
#2  0x00007ffff7a422b4 in ?? () from /usr/lib64/libtorrent-rasterbar.so.8
#3  0x0000000000464354 in boost::asio::detail::boost_asio_detail_posix_thread_function (arg=0x6dd170)
    at /usr/include/boost/asio/detail/impl/posix_thread.ipp:64
#4  0x00007ffff7494204 in start_thread () from /lib64/libpthread.so.0
#5  0x00007ffff67b178d in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7ffff7fd2780 (LWP 22341)):
#0  0x00007ffff749854f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff7ab3bc3 in libtorrent::torrent_handle::status(unsigned int) const ()
    from /usr/lib64/libtorrent-rasterbar.so.8
#2  0x0000000000480e2f in simple_torrent_client (
    path="/home/user/Downloads/test.torrent")
    at /home/user/dev/test_torrent/torrent.cpp:443
#3  0x000000000046f1d3 in main (argc=1, argv=0x7fffffffdac8) at /home/user/dev/test_torrent/main.cpp:66
```

**Comment by arvidn** [comment]
it doesn't look like the main thread is running anymore. I was expecting to see the main thread being stuck in torrent::status(). My new theory is that it fails with an exception, which is caught by the top level that session thread, that runs the main loop. Every time it catches an exception, it resets the io_service and runs it again. It appears as if it immediately fails again. One strange thing is that frame 2 in thread 2 does not have a symbol (that's where I was expecting to see session_impl).

Would you be able to verify this by, before starting the program, on the gdb prompt, run:

`catch throw`

if that fails, try:

`break __cxa_throw`

If it breaks, please capture the stack

**Comment by evgenyzinoviev** [comment]
Nope, it didn't work, no exceptions apparently. :(

**Comment by arvidn** [comment]
do you see any log message in gdb that a thread exited perhaps? Normally I would expect to see at least 4 threads. the main application thread, the network thread, boost.asio starts two threads for DNS lookups and a disk thread. It's possible asio only starts a single thread for hostname lookups, either way, there's at least one thread missing. The lack of symbols in the libtorrent library on the other threads makes it a bit hard to tell which threads are missing. Maybe there's a way to have gdb stop on thread exit.

**Comment by arvidn** [comment]
perhaps you could try this?

`catch thread_exit`

**Comment by evgenyzinoviev** [comment]
OK, I'll try in few minutes. But I didn't see any thread exit messages... Maybe I have to rebuild libtorrent with some debugging option enabled, it it has one?

**Comment by evgenyzinoviev** [comment]
1)
```
(gdb) catch thread_exit
Undefined catch command: "thread_exit".  Try "help catch".
```
2) Just noticed that backtrace for Thread 2 is different from time to time. I don't know it this matters. For example, now it was:
```
Thread 2 (Thread 0x7ffff51fc700 (LWP 10599)):
#0  0x00007ffff74960ab in pthread_mutex_lock () from /lib64/libpthread.so.0
#1  0x0000000000470300 in boost::asio::detail::posix_mutex::lock (this=0x6da090)
    at /usr/include/boost/asio/detail/posix_mutex.hpp:52
#2  0x000000000047243a in boost::asio::detail::scoped_lock<boost::asio::detail::posix_mutex>::scoped_lock (
    this=0x7ffff51fbd90, m=...) at /usr/include/boost/asio/detail/scoped_lock.hpp:46
#3  0x000000000046b4cc in boost::asio::detail::task_io_service::stop (this=0x6da060)
    at /usr/include/boost/asio/detail/impl/task_io_service.ipp:233
#4  0x000000000046af7c in boost::asio::detail::task_io_service::run (this=0x6da060, ec=...)
    at /usr/include/boost/asio/detail/impl/task_io_service.ipp:138
#5  0x0000000000460319 in boost::asio::io_service::run (this=0x6d8b18, ec=...)
    at /usr/include/boost/asio/impl/io_service.ipp:66
#6  0x00007ffff7a422ac in ?? () from /usr/lib64/libtorrent-rasterbar.so.8
#7  0x0000000000464354 in boost::asio::detail::boost_asio_detail_posix_thread_function (arg=0x6dd170)
    at /usr/include/boost/asio/detail/impl/posix_thread.ipp:64
#8  0x00007ffff7494204 in start_thread () from /lib64/libpthread.so.0
#9  0x00007ffff67b178d in clone () from /lib64/libc.so.6
```

3) no exit messages while running:
```
 $ gdb /home/user/.clion11/system/cmake/generated/45aacc61/45aacc61/Debug0/test_torrent 
GNU gdb (Gentoo 7.9.1 vanilla) 7.9.1
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/user/.clion11/system/cmake/generated/45aacc61/45aacc61/Debug0/test_torrent...done.
(gdb) run
Starting program: /home/user/.clion11/system/cmake/generated/45aacc61/45aacc61/Debug0/test_torrent 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) checking_resume_data
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
0.00% (down: 0.0 kb/s, up: 0.0 kb/s, peers: 0, paused: 1, auto_managed: 0, qp: 0) downloading
^C[New Thread 0x7ffff51fc700 (LWP 10843)]
[New Thread 0x7ffff59fd700 (LWP 10842)]

Program received signal SIGINT, Interrupt.
0x00007ffff749854f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
(gdb) 
```

**Comment by evgenyzinoviev** [comment]
No more ideas what could cause this?

**Comment by evgenyzinoviev** [comment]
Just tried it on my OS X machine, works fine, so that is something related to linux or my specific environment, or probably boost..
I'll try to build it on another linux env and see what happens.

**Comment by evgenyzinoviev** [comment]
Tried it on another machine with ubuntu 14.04 (previous was gentoo), 64-bit, boost 1.55, libtorrent 1.0.6 - same thing, freezes after few seconds...

**Comment by arvidn** [comment]
btw. are you saying that if you don't pause the torrent you don't see any problems?

**Comment by evgenyzinoviev** [comment]
Yes. It is downloading ok.

**Comment by Falcosc** [comment]
@evgenyzinoviev could you still reproduce it with 1.0.10 or even the head of 1.0?
